algorithms/linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,498 @@
+use crate::optics::errors::Result;
+use crate::optics::hyperparameters::OpticsHyperParams;
+use float_ord::FloatOrd;
```

Comment:
`FloatOrd` ordering semantics don't handle NaN at all, since NaN is listed as the largest and smallest value in the docs. Assuming that you don't actually want NaNs in your input at all, the `noisy_float` crate is the better choice. The type `noisy_float::NoisyFloat` implements `Ord` and also panics when you attempt to construct it with NaN in debug mode, so it does a better job of upholding the implicit assumption of no NaNs.